packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts

```diff
@@ -0,0 +1,179 @@
+
+type LegacyRuntime = {
+  value: 'nodejs' | 'python';
+  name: 'NodeJS' | 'Python';
+  layerExecutablePath: string;
+  cloudTemplateValue: string;
+};
+
+type LegacyVersionMap = {
+  [key: string]: {
+    permissions: LegacyPermission[];
+    hash?: string;
+  };
+};
+
+const enum LegacyState {
+  NOT_LEGACY,
+  MULTI_ENV_LEGACY,
+  SINGLE_ENV_LEGACY,
+}
+
+const layerVersionMapKey = 'layerVersionMap';
+
+export async function migrateLegacyLayer(context: $TSContext, layerName: string): Promise<boolean> {
+  const layerDirPath = getLayerDirPath(layerName);
+  const legacyState = isLegacyLayer(layerName, layerDirPath);
+
+  if (legacyState === LegacyState.NOT_LEGACY) {
+    return false;
+  }
+
+  const runtimecloudTemplateValues: string[] = [];
+  let layerConfiguration: { permissions: LayerPermission[]; runtimes: LayerRuntime[]; nonMultiEnv?: boolean } = {
+    permissions: undefined,
+    runtimes: undefined,
+  };
+  let legacyRuntimeArray: LegacyRuntime[];
+  let layerVersionMap: LegacyVersionMap;
+
+  if (legacyState === LegacyState.MULTI_ENV_LEGACY) {
+    legacyRuntimeArray = JSONUtilities.readJson<LegacyRuntime[]>(path.join(layerDirPath, LegacyFilename.layerRuntimes));
+    layerVersionMap = _.get(stateManager.getMeta(), [categoryName, layerName, layerVersionMapKey], {});
+  } else {
+    ({ layerVersionMap, runtimes: legacyRuntimeArray } = JSONUtilities.readJson<{
+      [layerVersionMapKey]: LegacyVersionMap;
+      runtimes: LegacyRuntime[];
+    }>(path.join(layerDirPath, LegacyFilename.layerParameters)));
+    layerConfiguration.nonMultiEnv = true;
+  }
+
+  legacyRuntimeArray
+    .map(legacyRuntime => legacyRuntime.cloudTemplateValue)
+    .forEach((cloudTemplateValue: string) => {
+      runtimecloudTemplateValues.push(cloudTemplateValue);
+    });
+
+  legacyRuntimeArray.map((runtime: LegacyRuntime) => delete runtime.cloudTemplateValue);
+  layerConfiguration.runtimes = legacyRuntimeArray;
+
+  await Promise.all(
+    layerConfiguration.runtimes.map(async runtime => {
+      if (runtime.value === 'nodejs') {
+        runtime.runtimePluginId = 'amplify-nodejs-function-runtime-provider';
+      } else if (runtime.value === 'python') {
+        runtime.runtimePluginId = 'amplify-python-function-runtime-provider';
+      }
+      const runtimePlugin = await loadPluginFromFactory(runtime.runtimePluginId, 'functionRuntimeContributorFactory', context);
+      const runtimeInfo = await runtimePlugin.contribute({ selection: runtime.value });
+      runtime.layerExecutablePath = runtimeInfo.runtime.layerExecutablePath;
+    }),
+  );
+
+  const layerVersions = Object.keys(layerVersionMap)
+    .map(version => parseInt(version, 10))
+    .sort((a, b) => b - a);
+
+  const permissions: LegacyPermission[] = layerVersionMap[`${_.first(layerVersions)}`]?.permissions;
+
+  if (permissions === undefined) {
+    context.print.warning('Unable to find layer permissions, falling back to default.');
```

Comment:
I think we need to provide more information here for better DX: layer name for example as if the project has multiple layers defined this does not give enough information to the customer.